src/scheduler/pbs_sched.cpp

```diff
@@ -1199,53 +1084,45 @@ main(int argc, char *argv[])
 
 	pthread_mutex_init(&cleanup_lock, &attr);
 
-	connect_servers();
+	sched_svr_init();
+	connect_svrpool();
 
 	for (go=1; go;) {
+		int i;
 
 		wait_for_cmds();
-		while (go && !ds_queue_is_empty(sched_cmds)) {
-			sched_cmd *cmd = static_cast<sched_cmd *>(ds_dequeue(sched_cmds));
-			sched_svrconn *sconn = NULL;
 
-			if (!cmd)
-				continue;
-
-			sconn = find_server(cmd->from_sock);
-			if (!sconn) {
-				/* shouldn't happen, but if it does just continue */
-				free_sched_cmd(cmd);
-				continue;
+		/* First walk through the qrun_list as this has high priority followed by the other commands */
+		for (i = 0; i < qrun_list_size; i++) { 
+			if (schedule_wrapper(&qrun_list[i], opt_no_restart) == 1) {
+				go = 0;
+				break;
 			}
+			free(qrun_list[i].jid);
+			qrun_list[i].jid = NULL;
+		}
 
-#ifdef PBS_UNDOLR_ENABLED
-			if (sigusr1_flag)
-				undolr();
-#endif
-			if (sigprocmask(SIG_BLOCK, &allsigs, &oldsigs) == -1)
-				log_err(errno, __func__, "sigprocmask(SIG_BLOCK)");
+		for (i = 0; go && (i < SCH_CMD_HIGH); i++) {
+			sched_cmd cmd;
 
-			/* Keep track of time to use in SIGSEGV handler */
-			now = time(NULL);
-			if (!opt_no_restart)
-				segv_last_time = now;
-#ifdef DEBUG
-			{
-				strftime(log_buffer, sizeof(log_buffer), "%Y-%m-%d %H:%M:%S", localtime(&now));
-				DBPRT(("%s Scheduler received command %d\n", log_buffer, sconn->cmd));
-			}
-#endif
+			if (sched_cmds[i] == 0)
+				continue;
 
-			if (schedule(sconn, cmd)) /* magic happens here */
-				go = 0;
-			else
-				send_cycle_end(sconn);
+			/* index itself is the command */
+			cmd.cmd = i;
 
-			free_sched_cmd(cmd);
+			/* jid is always NULL since this list does not contain SCHEDULE_AJOB commands */
+			cmd.jid = NULL;
 
-			if (sigprocmask(SIG_SETMASK, &oldsigs, NULL) == -1)
-				log_err(errno, __func__, "sigprocmask(SIG_SETMASK)");
+			/* clear the entry of sched_cmds[i] as we are going to process this command now */
+			sched_cmds[i] = 0;
```

Comment:
Since '0' is a valid command, I'd set it to -1 and check for it to be not -1 before calling the wrapper.

Reply:
The command is actually stored as an index, the value at this index just indicates whether a particular command is received or not. So even though if zero is valid command we still honor it so it might look like sched_cmds[SCH_SCHEDULE_NULL] = 1 (SCH_SCHEDULE_NULL = 0) if we receive SCH_SCHEDULE_NULL or
sched_cmds[SCH_SCHEDULE_NULL] = 0 if we don't receive it

I feel 0 or 1 is more readable here and it is not causing any problem in my testing. So I want to leave this as it. Hope this is fine.

We can actually initialize all the elements of the array to -1 in the beginning once also, But this needs an extra loop. As of now sched_cmds[] is global so by default it is initialized to zero which is what is needed for us.